### DISCUSSION TITLE
SQLiteを動かすためにするべきことについて

### DISCUSSION THREAD

**Discussion by keisuke713** [nuta/microkernel-book#11] [Help] [answered]
### 概要

[こちら](https://github.com/nuta/microkernel-book/blob/main/IDEAS.md#sqlite%E3%82%B5%E3%83%BC%E3%83%90)に書かれているSQLiteの移植をやりたいのですが、ソースコードをコピーしてきた後に何をすべきか明確にしたいです。
[公式](https://www.sqlite.org/download.html)のページから `sqlite-amalgamation-3420000.zip` の内容をコピーして `make run` すると  `servers/sqlite3/shell.c:116:10: fatal error: 'stdlib.h' file not found ` などと出るので一旦 `shell.c` `sqlite3.c` 内でインクルードされている標準ライブラリを全てコメントアウトしましたが、そうすると今度は以下のように標準ライブラリの実装を参照している箇所で(当たり前ですが)定義されていないとエラーが出ます。
```
servers/sqlite3/shell.c:238:11: error: unknown type name 'FILE'
   extern FILE *popen(const char*,const char*);
          ^
servers/sqlite3/shell.c:239:22: error: unknown type name 'FILE'
   extern int pclose(FILE*);
                     ^
servers/sqlite3/shell.c:299:10: error: call to undeclared function 'strcmp'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
  return strcmp(a,b);
```

1.ここまでの僕の手順(ソースコードをコピーして標準ライブラリのインクルードをコメントアウトする)は合っていますでしょうか？
2.次の手順としては標準ライブラリで定義されていたものを全てHinaOS内で実装してそのファイルをインクルードしていくのかな？と考えているのですが、合っていますでしょうか？

### 開発環境

```bash
OS:      macOS-13.2.1-x86_64-i386-64bit
PYTHON3: Python 3.10.10
CC:      Homebrew clang version 16.0.6 Target: x86_64-apple-darwin22.3.0 Thread model: posix InstalledDir: /usr/local//opt/llvm/bin
LD:      Homebrew LLD 16.0.6 (compatible with GNU linkers)
QEMU:    QEMU emulator version 8.0.3 Copyright (c) 2003-2022 Fabrice Bellard and the QEMU Project developers
OBJCOPY: llvm-objcopy, compatible with GNU objcopy Homebrew LLVM version 16.0.6   Optimized build.
MAKE:    GNU Make 3.81 Copyright (C) 2006  Free Software Foundation, Inc. This is free software; see the source for copying conditions. There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  This program built for i386-apple-darwin11.3.0
```
```


### ソースコードのURL

https://github.com/keisuke713/microkernel-book

**Comment by nuta** [comment]
まず初めに「sqliteの移植」というのは「C標準ライブラリの用意」「sqliteが動くようにビルド設定を調整する」「sqliteが利用するOS依存部分の実装」の3パートに分かれます。今取り組まれているのは最初のC標準ライブラリの用意パートですね。

> 1.ここまでの僕の手順(ソースコードをコピーして標準ライブラリのインクルードをコメントアウトする)は合っていますでしょうか？

`sqlite3.c` 本体を書き換えていくのは大変なので、コメントアウトしていくのではなく標準ライブラリ自体を用意する方針がよいと思います。自作OSのような環境だと、newlibやpdclibといった移植しやすい実装が使えます。

なお、 `shell.c` はWindows/UNIX向けに書かれているオプショナルな物なので、まずは `sqlite3.c` (SQLite本体) のみを移植するのがおすすめです。

> 2.次の手順としては標準ライブラリで定義されていたものを全てHinaOS内で実装してそのファイルをインクルードしていくのかな？と考えているのですが、合っていますでしょうか？

C標準ライブラリが必要とするもの ([newlibの例](https://sourceware.org/newlib/)) を実装することになります。`open` などOSの手助けが必要な低レベルのものです。ただしシンプルな`SELECT`文であれば `strcspn` と `strrchr` しか利用しないようなので、インクルード部分だけ上手くいくようにすればそのまま動くと思います。

最後に、[`sqlite3`ブランチ](https://github.com/nuta/microkernel-book/tree/sqlite3) にpdclibを利用した実装例を置いておきました。コードからは把握しずらいSQLite移植の落とし穴がいくつかあるので、ブログかどこかにまとめるつもりです。書き終わり次第ここでも共有します。

**Reply by keisuke713** [reply]
手順を示していただきありがとうございます。
方針が明確ではなかったのでとても助かります。

> sqlite3.c 本体を書き換えていくのは大変なので、コメントアウトしていくのではなく標準ライブラリ自体を用意する方針がよいと思います。

僕の認識ですと `#include <>` は標準ライブラリを読み込むときに使用すると思うのですが、nutaさんの実装例を拝見させていただいたところ、なぜ `sqlite3.c` 内で変わらず `#include <stdlib.h>` などと読み込むことが可能なのでしょうか？
`sqlite3.c` が参照している `stdlib.h` は新しく用意されたもの( `pdclib`配下のファイルたち )なので `#include "stdlib.h"` になるのでは？と思いました。

どこで `pdclib`配下のファイルたちを標準ライブラリとして読み込めるように設定されているのでしょうか。
最初は `servers/sqlite/build.mk` にコンパイラオプションのようなものを書かれているのかな？とも思いましたが、[こちら](https://github.com/nuta/microkernel-book/commit/298d72b978dbc34b9092d110499d2fb3a2cef979#diff-d7c775c653721bad7e3074e884845b1e2fa6b1f46b532edaa97db0d87caf3912) のように `build.mk` を書き足しても変わらず下記のようなエラーが出たので気になった次第です。
```
servers/sqlite/sqlite3.c:345:10: error: 'stdarg.h' file not found with <angled> include; use "quotes" instead
#include <stdarg.h>     /* Needed for the definition of va_list */
         ^~~~~~~~~~
         "stdarg.h"
servers/sqlite/sqlite3.c:14292:10: fatal error: 'stdio.h' file not found
#include <stdio.h>
         ^~~~~~~~~
2 errors generated.
make: *** [build/servers/sqlite/sqlite3.o] Error 1
```

**Reply by nuta** [reply, marked as answer]
`stdarg.h` と `stddef.h` に関してはPDCLibが提供していないので、`servers/sqlite3` 配下に別に追加しています。次の行で それらをインクルードできるようにしています。

https://github.com/nuta/microkernel-book/blob/2d284bceb79e4fb35c56b109c8b05ca2ec2eb137/servers/sqlite3/build.mk#L9

また、移植例の解説を[こちら](https://github.com/nuta/microkernel-book/tree/sqlite3#sqlite3-on-hinaos)に追加しました。ご参考になれば幸いです。

**Reply by keisuke713** [reply]
READMEにまでまとめてくださり、大変ありがとうございます。
nutaさんのブランチを参考に書き込み処理の実装を進めていこうと思います。

また詰まったら質問させてください 🙏

**Comment by keisuke713** [comment]
@nuta
